types/urijs/index.d.ts

```diff
@@ -118,17 +118,15 @@ declare namespace URI {
         preventInvalidHostname: boolean;
     }
 
-    interface QueryDataMap {
-        [key: string]: string | null | Array<string | null>;
-    }
+    type QueryDataMap = Record<string, any>;
```

Comment:
This is likely less safe than before. `any` can take literally anything, including objects, `Map`s, and a lot of things we shouldn't allow. What's wrong with taking your list (assuming you double checked it is correct): `boolean | number | string | null | undefined | Array<boolean | number | string | null | undefined>`. 

Also to be correct we should make it `Partial<Record<....>>` because of course it is a partial. (Non-partial `Record<string>`s almost never make any sense.)

Reply:
QueryDataMaps can take literally anything including objects, `Map`s, and a lot of things

Reply:
> Non-partial Record<string>s almost never make any sense.

These are parameters for a URL

Reply:
> What's wrong with taking your list?

The issue that I see is that it would be incomplete, because urijs queries _can_ support any object, even if it seems like it shouldn't. https://github.com/medialize/URI.js/blob/gh-pages/src/URI.js#L782-L811

```js
 map = new Map()
 map.set('a', 'b')
 uri('github.com').addQuery({'map':map}).toString() // => 'github.com?map=%5Bobject+Map%5D'
```